runtime/sema/resource_invalidations.go

```diff
@@ -37,33 +76,57 @@ func (ris ResourceInvalidations) All() (result []ResourceInvalidation) {
 	return
 }
 
-func (ris ResourceInvalidations) Include(invalidation ResourceInvalidation) bool {
-	return ris.invalidations.Include(ResourceInvalidationEntry{
-		ResourceInvalidation: invalidation,
-	})
+// Add adds the given resource invalidation to this set.
+//
+func (ris *ResourceInvalidations) Add(invalidation ResourceInvalidation) {
+	if ris.Contains(invalidation) {
+		return
+	}
+	if ris.invalidations == nil {
+		ris.invalidations = map[ResourceInvalidation]struct{}{}
+	}
+	ris.invalidations[invalidation] = struct{}{}
 }
 
-func (ris ResourceInvalidations) Insert(invalidation ResourceInvalidation) ResourceInvalidations {
-	entry := ResourceInvalidationEntry{invalidation}
-	newInvalidations := ris.invalidations.Insert(entry)
-	return ResourceInvalidations{newInvalidations}
+// Delete removes the given resource invalidation from this set.
+//
+func (ris *ResourceInvalidations) Delete(invalidation ResourceInvalidation) {
+	if ris.invalidations == nil {
+		return
+	}
+	delete(ris.invalidations, invalidation)
```

Comment:
should it need to check and delete in the `ris.Parent` also?

There is a chance that `ris.Delete(x)` followed by a `ris.Contains(x)` might still return `true`

Reply:
Good catch/question! The semantics are a bit weird here.

In general, all functions should only ever operate on the current map/set (i.e. child), never the parent. The idea is that operations in a child environment, e.g. branches of an if-statement, don't effect the parent environment.

To properly implement this for deletion, we could mark the entry as deleted in the child (sort of like a "tombstone").

However, the delete function here only exists and is only used `Resources.RemoveTemporaryInvalidation`, which is called after a temporary invalidation was added to the resources map before. So all that is needed for this function is the current semantics (it's OK if the element was temporarily added before in the parent, then temporarily added in the child, then removed from the child, ... and now contains still returns true).

I don't think we have to implement proper deletion as described above, but rather, I'll rename the function and make it clear that it only applies to this child, and no deletion is performed for the parent.